xcm/pallet-xcm/src/lib.rs

```diff
@@ -1306,22 +1310,24 @@ impl<T: Config> Pallet<T> {
 		interior: impl Into<Junctions>,
 		dest: impl Into<MultiLocation>,
 		mut message: Xcm<()>,
-	) -> Result<(), SendError> {
+	) -> Result<XcmHash, SendError> {
 		let interior = interior.into();
 		let dest = dest.into();
-		let maybe_fee_payer = if interior != Junctions::Here {
+		let maybe_fee_payer: Option<MultiLocation> = if interior != Junctions::Here {
 			message.0.insert(0, DescendOrigin(interior.clone()));
 			Some(interior.into())
 		} else {
 			None
 		};
 		log::trace!(target: "xcm::send_xcm", "dest: {:?}, message: {:?}", &dest, &message);
+		let message_hash =
+			VersionedXcm::from(message.clone()).using_encoded(sp_io::hashing::blake2_256);
```

Comment:
This is a heavy operation. If `charge_fees` needs to be passed in the context, then refactor it so that `validate_send` returns the message hash rather than `deliver`.